src/Functions/toTimezone.cpp

```diff
@@ -99,7 +99,14 @@ class ToTimeZoneOverloadResolver : public IFunctionOverloadResolver
             throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
                 "Should be DateTime or DateTime64", arguments[0].type->getName(), getName());
 
-        String time_zone_name = extractTimeZoneNameFromFunctionArguments(arguments, 1, 0);
+        String time_zone_name;
+        if (arguments[1].column)
+            time_zone_name = extractTimeZoneNameFromColumn(*arguments[1].column);
+
+        if (time_zone_name.empty())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
+                "Should be const String", arguments[1].type->getName(), getName());
```

Comment:
`extractTimeZoneNameFromColumn` throws the exception about illegal type:
https://github.com/ClickHouse/ClickHouse/blob/70d1adfe4b7e437a62c59ca44a51fbd186ec7584/src/Functions/extractTimeZoneFromFunctionArguments.cpp#L20-L29

It cannot return an empty string unless an empty timezone name was passed as the second argument

Reply:
The question is: why do we need the check for `time_zone_name.empty()` hare? Can we just remove the duplicate check from `ToTimeZoneOverloadResolver::getReturnTypeImpl`?

Reply:
Yes, we can remove the check from `ToTimeZoneOverloadResolver::getReturnTypeImpl`. But for doing so we would need to change `extractTimeZoneNameFromColumn` to also raise exception when column is not const.

You can see how `checkAndGetColumnConst` don't return null when column is not const but empty:

https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionHelpers.h#L35-L38

Reply:
Ok, got it. The thing is that `arguments[1].column` is not declared when column is not const. So I can't call `extractTimeZoneNameFromColumn` straight away because `*arguments[1].column` will raise null pointer error. That's why I've moved the check to `ToTimeZoneOverloadResolver::getReturnTypeImpl`

The original code was using `extractTimeZoneNameFromFunctionArguments`, doesn't have this issue but it tries to get time zone from the DateTime is can't get it from the time zone param, something it think is not right.

Any suggestion here?

Reply:
What do you think about changing `extractTimeZoneNameFromFunctionArguments` to something like this? I'm not sure about side effects of the change

```diff
--- a/src/Functions/extractTimeZoneFromFunctionArguments.cpp
+++ b/src/Functions/extractTimeZoneFromFunctionArguments.cpp
@@ -33,8 +33,13 @@ std::string extractTimeZoneNameFromColumn(const IColumn & column)
 std::string extractTimeZoneNameFromFunctionArguments(const ColumnsWithTypeAndName & arguments, size_t time_zone_arg_num, size_t datetime_arg_num)
 {
     /// Explicit time zone may be passed in last argument.
-    if (arguments.size() == time_zone_arg_num + 1 && arguments[time_zone_arg_num].column)
+    if (arguments.size() == time_zone_arg_num + 1)
     {
+        if (!arguments[time_zone_arg_num].column) {
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                            "Illegal column {} of time zone argument of function, must be constant string",
+                            arguments[time_zone_arg_num].name);
+        }
         return extractTimeZoneNameFromColumn(*arguments[time_zone_arg_num].column);
     }
     else
```

Reply:
I'm getting the following error when calling `extractTimeZoneNameFromColumn(*arguments[1].column)` without any further check.

```
[jordi-mbp] 2023.04.11 17:16:25.001009 [ 2243297 ] <Fatal> BaseDaemon: ########################################
[jordi-mbp] 2023.04.11 17:16:25.001101 [ 2243297 ] <Fatal> BaseDaemon: (version 23.4.1.1, build id: ) (from thread 2241562) (query_id: eb208073-09db-4fa9-8c05-460ea2ca0874) (query: SELECT materialize('Asia/Kolkata') t, toTimeZone(toDateTime('2017-11-05 08:07:47', 'Asia/Istanbul'), t);) Received signal Segmentation fault: 11 (11)
[jordi-mbp] 2023.04.11 17:16:25.001134 [ 2243297 ] <Fatal> BaseDaemon: Address: NULL pointer. Access: . Attempted access has violated the permissions assigned to the memory area.
[jordi-mbp] 2023.04.11 17:16:25.001155 [ 2243297 ] <Fatal> BaseDaemon: Stack trace: 0x109a46044
[jordi-mbp] 2023.04.11 17:16:25.001175 [ 2243297 ] <Fatal> BaseDaemon: 0. 0x109a46044
```

Reply:
It can be nullptr during `ExpressionAnalysis` if the column is not const.

The callstack is something like:
```
1. ./build/./src/Common/StackTrace.h:0: DB::extractTimeZoneNameFromFunctionArguments(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, unsigned long, unsigned long) @ 0x15ecfcde in /mnt/ch/ClickHouse/build_default/programs/clickhouse
2. DB::(anonymous namespace)::ToTimeZoneOverloadResolver::getReturnTypeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&) const @ 0x10d18043 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
3. ./build/./src/Functions/IFunction.cpp:481: DB::IFunctionOverloadResolver::getReturnTypeWithoutLowCardinality(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&) const @ 0x15ec41b4 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
4. ./build/./src/Functions/IFunction.cpp:427: DB::IFunctionOverloadResolver::getReturnType(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&) const @ 0x15ec3ec1 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
5. ./build/./src/Functions/IFunction.cpp:442: DB::IFunctionOverloadResolver::build(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&) const @ 0x15ec47e2 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
6. ./build/./src/Interpreters/ActionsDAG.cpp:198: DB::ActionsDAG::addFunction(std::shared_ptr<DB::IFunctionOverloadResolver> const&, std::vector<DB::ActionsDAG::Node const*, std::allocator<DB::ActionsDAG::Node const*>>, String) @ 0x16634d6d in /mnt/ch/ClickHouse/build_default/programs/clickhouse
7. ./build/./src/Interpreters/ActionsVisitor.cpp:0: DB::ScopeStack::addFunction(std::shared_ptr<DB::IFunctionOverloadResolver> const&, std::vector<String, std::allocator<String>> const&, String) @ 0x1680122d in /mnt/ch/ClickHouse/build_default/programs/clickhouse
8. ./build/./contrib/llvm-project/libcxx/include/string:1499: DB::ActionsMatcher::Data::addFunction(std::shared_ptr<DB::IFunctionOverloadResolver> const&, std::vector<String, std::allocator<String>> const&, String) @ 0x1680e07d in /mnt/ch/ClickHouse/build_default/programs/clickhouse
9. ./build/./contrib/llvm-project/libcxx/include/string:1499: DB::ActionsMatcher::visit(DB::ASTFunction const&, std::shared_ptr<DB::IAST> const&, DB::ActionsMatcher::Data&) @ 0x16803504 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
10. ./build/./src/Interpreters/ActionsVisitor.cpp:830: DB::ActionsMatcher::visit(DB::ASTExpressionList&, std::shared_ptr<DB::IAST> const&, DB::ActionsMatcher::Data&) @ 0x1680a249 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
11. ./build/./src/Interpreters/InDepthNodeVisitor.h:0: DB::InDepthNodeVisitor<DB::ActionsMatcher, true, false, std::shared_ptr<DB::IAST> const>::doVisit(std::shared_ptr<DB::IAST> const&) @ 0x167f9023 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
12. ./build/./src/Interpreters/InDepthNodeVisitor.h:0: DB::ExpressionAnalyzer::getRootActions(std::shared_ptr<DB::IAST> const&, bool, std::shared_ptr<DB::ActionsDAG>&, bool) @ 0x167d967e in /mnt/ch/ClickHouse/build_default/programs/clickhouse
13. ./build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::SelectQueryExpressionAnalyzer::appendSelect(DB::ExpressionActionsChain&, bool) @ 0x167e4f35 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
14. ./build/./src/Interpreters/ExpressionAnalyzer.cpp:2086: DB::ExpressionAnalysisResult::ExpressionAnalysisResult(DB::SelectQueryExpressionAnalyzer&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, bool, bool, bool, std::shared_ptr<DB::FilterDAGInfo> const&, std::shared_ptr<DB::FilterDAGInfo> const&, DB::Block const&) @ 0x167eac03 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
15. ./build/./src/Interpreters/InterpreterSelectQuery.cpp:931: DB::InterpreterSelectQuery::getSampleBlockImpl() @ 0x17097185 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
16. ./build/./src/Interpreters/InterpreterSelectQuery.cpp:764: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>)::$_4::operator()(bool) const @ 0x1709109f in /mnt/ch/ClickHouse/build_default/programs/clickhouse
17. ./build/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:815: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x1708c102 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
18. ./build/./contrib/llvm-project/libcxx/include/optional:260: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x170895c2 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
19. ./build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:257: DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::shared_ptr<DB::IAST> const&, std::vector<String, std::allocator<String>> const&) @ 0x1711b5fa in /mnt/ch/ClickHouse/build_default/programs/clickhouse
20. ./build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:152: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x17119498 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
21. ./build/./contrib/llvm-project/libcxx/include/vector:434: std::__unique_if<DB::InterpreterSelectWithUnionQuery>::__unique_single std::make_unique[abi:v15000]<DB::InterpreterSelectWithUnionQuery, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&>(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&) @ 0x17051621 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
22. ./build/./src/Interpreters/InterpreterFactory.cpp:0: DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x17050815 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
23. ./build/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x173e51e4 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
24. ./build/./src/Interpreters/executeQuery.cpp:1166: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x173e36a8 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
25. ./build/./src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x180f6daf in /mnt/ch/ClickHouse/build_default/programs/clickhouse
26. ./build/./src/Server/TCPHandler.cpp:2040: DB::TCPHandler::run() @ 0x18107a0a in /mnt/ch/ClickHouse/build_default/programs/clickhouse
27. ./build/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x1aa46f47 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
28. ./build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: Poco::Net::TCPServerDispatcher::run() @ 0x1aa4743a in /mnt/ch/ClickHouse/build_default/programs/clickhouse
29. ./build/./base/poco/Foundation/src/ThreadPool.cpp:202: Poco::PooledThread::run() @ 0x1abc777b in /mnt/ch/ClickHouse/build_default/programs/clickhouse
30. ./build/./base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::ThreadImpl::runnableEntry(void*) @ 0x1abc5096 in /mnt/ch/ClickHouse/build_default/programs/clickhouse
31. ? @ 0x7f9fbbfadbb5 in ?
32. ? @ 0x7f9fbc02fd90 in ?
```

In fact I'd say that during `getReturnType` it's more common to have it being a nullptr than not as most ActionsDAG::Node won't declare it. I think `ActionsDAG::addInput(std::string name, DataTypePtr type)` is what's used for identifiers, vs `ActionsMatcher::visit(const ASTLiteral & literal`).

In any case, the handling of the timezone parameter seems all over the place as each individual function is doing the checks manually. In a quick look I see  that `now64` has the same problem (and should be solved by the patch) or that many of the conversions check for extractTimeZoneNameFromFunctionArguments().empty and throw too

Reply:
Hm, sorry, you are right. I was really confused by the fact that it's nullptr during query analysis for non-const columns. 

Yes, it makes sense to update `extractTimeZoneNameFromFunctionArguments`/`extractTimeZoneNameFromColumn` because we have exactly the same issue in other functions relying on `extractTimeZoneNameFromFunctionArguments`.